File: scripts/genealogy/generate_landau.py
#!/usr/bin/env python3
"""Stage 2: Landau Mean-Field Free Energy — equilibrium landscape.

Thin runner: imports all math from phosphene.genealogy.formulas.
"""
import json
import os

import numpy as np
from phosphene.genealogy.formulas import landau_free_energy, landau_equilibrium

SEED = 42
OUTPUT_DIR = os.path.join(os.path.dirname(__file__), "..", "..", "data", "genealogy")


def run_landau_simulation(n_points=100, a_range=(-1.0, 1.0), b=1.0, h=0.0):
    np.random.seed(SEED)
    a_vals = np.linspace(a_range[0], a_range[1], n_points)
    M_grid = np.linspace(-1.5, 1.5, 200)
    M_eq = np.zeros(n_points)
    F_min = np.zeros(n_points)

    for i, a in enumerate(a_vals):
        M_eq[i], F_min[i] = landau_equilibrium(M_grid, a, b, h)

    return {
        "stage": "landau",
        "formula": "F(M) = a·M² + b·M⁴ + h·M",
        "parameters": {"a_range": list(a_range), "b": b, "h": h},
        "data": {
            "a_values": a_vals.tolist(),
            "M_equilibrium": M_eq.tolist(),
            "F_min": F_min.tolist(),
        },
        "metadata": {
            "seed": SEED,
            "code": "phosphene.genealogy.formulas.landau_free_energy + landau_equilibrium",
        },
    }


def main():
    os.makedirs(OUTPUT_DIR, exist_ok=True)
    result = run_landau_simulation()
    out_path = os.path.join(OUTPUT_DIR, "landau_simulation.json")
    with open(out_path, "w") as f:
        json.dump(result, f, indent=2)
    print(f"Landau simulation complete → {out_path}")
    print(f"  Critical point (a=0): M_eq = {result['data']['M_equilibrium'][50]:.3f}")
    print(f"  Below Tc (a=-0.5): M_eq = {result['data']['M_equilibrium'][25]:.3f}")
    print(f"  Seed: {SEED}")


if __name__ == "__main__":
    main()

File: inst/genealogy/generate_landau.R
#' Landau Mean-Field Free Energy — Stage 2 of the Genealogy
#'
#' F(M) = a*M^2 + b*M^4 + h*M
#' Equilibrium: minimize F over M grid.
#'
#' @param seed Integer. Default 42.
#' @param n_points Integer. Number of M values. Default 100.
#' @param a_range Numeric. Range of 'a'. Default c(-1, 1).
#' @param b Numeric. Quartic coefficient. Default 1.0.
#' @param h Numeric. External field. Default 0.0.
#'
#' @return List with data frame: a, M_eq, F_min, T_norm, stage

generate_landau <- function(seed = 42L, n_points = 100L,
                            a_range = c(-1, 1), b = 1.0, h = 0.0) {
  withr::with_seed(seed, {
    a_vals <- seq(a_range[1], a_range[2], length.out = n_points)
    M_grid <- seq(-1.5, 1.5, length.out = 200)
    M_eq <- numeric(n_points)
    F_min <- numeric(n_points)

    for (i in 1:n_points) {
      a <- a_vals[i]
      F <- a * M_grid^2 + b * M_grid^4 + h * M_grid
      eq_idx <- which.min(F)
      M_eq[i] <- M_grid[eq_idx]
      F_min[i] <- F[eq_idx]
    }

    T_norm <- a_vals + 1

    data <- data.frame(
      a = a_vals, M_eq = M_eq, F_min = F_min,
      T_norm = T_norm, b = b, h = h, stage = "landau"
    )

    list(
      values = list(n_points = n_points, b = b, h = h),
      metadata = list(
        seed = seed, data = data,
        params = list(n_points = n_points, a_range = a_range, b = b, h = h),
        generator = "generate_landau", converged = TRUE
      )
    )
  })
}
